io/fasta/fasta.go

```diff
@@ -66,14 +70,181 @@ type Fasta struct {
+//     In this case the Fasta up to that point is returned with an EOF error.
+//
+// It is worth noting the amount of bytes read are always right up to before
+// the next fasta starts which means this function can effectively be used
+// to index where fastas start in a file or string.
+func (p *Parser) ParseNext() (Fasta, int64, error) {
+	if _, err := p.rd.Peek(1); err != nil {
+		// Early return on error. Probably will be EOF.
+		return Fasta{}, 0, err
+	}
+	// Initialization of parser state variables.
+	var (
+		// Parser looks for a line starting with '>' (U+003E)
+		// that contains the next fasta sequence name.
+		lookingForName = true
+		seqName        string
+		sequence, line []byte
+		err            error
+		totalRead      int64
+	)
+
+	// parse loop begins here.
+	for {
+		line, err = p.rd.ReadSlice('\n')
+		isSkippable := len(line) == 0 || line[0] == ';' // OR short circuits so no panic here.
+		totalRead += int64(len(line))
+		p.line++
+
+		// More general case of error handling.
+		if err != nil {
+			isEOF := errors.Is(err, io.EOF)
+			if isSkippable {
+				if isEOF {
+					// got EOF on a empty or commented line.
+					err = nil
+				}
+				break
+			} else if errors.Is(err, bufio.ErrBufferFull) {
+				// Buffer size too small to read fasta line.
+				return Fasta{}, totalRead, fmt.Errorf("line %d too large for buffer, use larger maxLineSize: %w", p.line+1, err)
+			} else if !isEOF {
+				return Fasta{}, totalRead, err // Unexpected error.
+			}
+
+			// So got to this point the line is probably OK, we will return a Fasta.
+			// with the EOF error.
+			sequence = append(sequence, line...)
+			break
+		}
 
-	var outputFastas []Fasta
-	for fasta := range fastas {
-		outputFastas = append(outputFastas, fasta)
+		line = line[:len(line)-1] // Exclude newline delimiter.
+		peek, _ := p.rd.Peek(1)
+		if !lookingForName && len(peek) == 1 && peek[0] == '>' {
+			// We are currently parsing a fasta and next line contains a new fasta.
+			// We handle this situation by appending current line to sequence if not a comment
+			// and ending the current fasta parsing.
+			if !isSkippable {
+				sequence = append(sequence, line...)
+			}
+			break
+		} else if isSkippable {
+			continue
+		}
+
+		if lookingForName {
+			if line[0] == '>' {
+				// We got the start of a fasta.
+				seqName = string(line[1:])
+				lookingForName = false
+			}
+			// This continue will also skip line if we are looking for name
+			// and the current line does not contain the name.
+			continue
+		}
+		// If we got to this point we are currently inside of the fasta
+		// sequence contents. We append line to what we found of sequence so far.
+		sequence = append(sequence, line...)
+	} // parse loop ends here.
+
+	// Parsing ended. Check for inconsistencies.
+	if lookingForName {
+		return Fasta{}, totalRead, fmt.Errorf("did not find fasta start '>', got to line %d: %w", p.line, err)
```

Comment:
I can't seem to reach this code while parsing with an empty fasta file. Is there a way to do this?

Reply:
Done!